File: 프로그래머스/lv2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.js
function solution(rows, columns, queries) {
    const arr = [...Array(rows)].map((v, r) =>
        [...Array(columns)].map((v, c) => r * columns + (c + 1))
    );
    return queries.reduce((acc, querie) => {
        const [x1, y1, x2, y2] = querie;
        const move = [x2 - x1, y2 - y1];
        const [dirX, dirY] = [
            [0, 1, 0, -1],
            [1, 0, -1, 0],
        ];

        let [x, y] = [x1 - 1, y1 - 1];
        let temp = arr[x][y],
            min = 9999999999;
        const swap = (a) => {
            const j = a;
            a = temp;
            temp = j;
            return a;
        };

        for (let i = 0; i < 4; ++i) {
            let cnt = move[(i + 1) % 2];

            while (cnt--) {
                (x += dirX[i]), (y += dirY[i]);
                arr[x][y] = swap(arr[x][y]);
                min = Math.min(min, arr[x][y]);
            }
        }
        acc.push(min);
        return acc;
    }, []);
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>

using namespace std;

enum EDIR
{
    RIGHT,
    DOWN,
    LEFT,
    UP,
    END
};

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    vector<vector<int>> board;
    const int move[EDIR::END][2] = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

    int cnt = 1;
    for (int i = 0; i < rows; i++)
    {
        std::vector<int> v;
        for (int j = 0; j < columns; j++) v.push_back(cnt++);
        board.push_back(v);
    }

    for (const auto& pos : queries)
    {
        int start[] = { pos[0], pos[1] };
        int temp = board[start[0] - 1][start[1] - 1], resMin = 99999;

        for (int dir = static_cast<int>(EDIR::RIGHT); dir < EDIR::END; ++dir)
        {
            while (true)
            {
                int nPos[] = { start[0] + move[dir][0], start[1] + move[dir][1] };

                if (nPos[0] < pos[0] || nPos[0] > pos[2] || nPos[1] < pos[1] || nPos[1] > pos[3])
                    break;

                for (int i = 0; i < 2; ++i)
                    start[i] += move[dir][i];

                swap(temp, board[nPos[0] - 1][nPos[1] - 1]);
                resMin = min(temp, resMin);
            }
        }
        answer.push_back(resMin);
    }

    return answer;
}